### ISSUE TITLE
format checkers execute before the instance is validated against the string type

### ISSUE THREAD

**Issue by filip-komarzyniec** [python-jsonschema/jsonschema#1519]
I believe the expected behaviour would be to first validate whether the instance conforms to the `string` type defined in the schema and only then run any further checks against it. 

With the current behaviour any custom validators have to ensure the type with their custom code rendering part of the jsonschema functionality completely useless. 

#### Minimal reproducible example

```python
schema = {
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "object",
  "properties": {
    "my-field": {
      "type": "string",
      "format": "email"
    },
  }
}
obj = {
  "my-field": 123
}

formatter = jsonschema.FormatChecker()

@formatter.checks("email", raises=ValueError)  # custom logic for email fields verification
def custom_email_validation(instance: str):
  if isinstance(instance, str):
    print("this is string")
  else:
    print("this is not string")
    raise RuntimeError("Some error in my custom logic due to unexpected error.")
  return False

```

Running the above gives the output:

```python-traceback
>> this is not string
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[28], line 1
----> 1 jsonschema.validate(obj, schema, format_checker=formatter)

File ~/kf_pipelines_components/lib/python3.12/site-packages/jsonschema/validators.py:1330, in validate(instance, schema, cls, *args, **kwargs)
   1328 cls.check_schema(schema)
   1329 validator = cls(schema, *args, **kwargs)
-> 1330 error = exceptions.best_match(validator.iter_errors(instance))
   1331 if error is not None:
   1332     raise error

File ~/kf_pipelines_components/lib/python3.12/site-packages/jsonschema/exceptions.py:479, in best_match(errors, key)
    440 def best_match(errors, key=relevance):
    441     """
    442     Try to find an error that appears to be the best match among given errors.
    443 
   (...)    477 
    478     """
--> 479     best = max(errors, key=key, default=None)
    480     if best is None:
    481         return

File ~/kf_pipelines_components/lib/python3.12/site-packages/jsonschema/validators.py:383, in create.<locals>.Validator.iter_errors(self, instance, _schema)
    381 for validator, k, v in validators:
    382     errors = validator(self, v, instance, _schema) or ()
--> 383     for error in errors:
    384         # set details if not already set by the called fn
    385         error._set(
    386             validator=k,
    387             validator_value=v,
   (...)    390             type_checker=self.TYPE_CHECKER,
    391         )
    392         if k not in {"if", "$ref"}:

File ~/kf_pipelines_components/lib/python3.12/site-packages/jsonschema/_keywords.py:296, in properties(validator, properties, instance, schema)
    294 for property, subschema in properties.items():
    295     if property in instance:
--> 296         yield from validator.descend(
    297             instance[property],
    298             subschema,
    299             path=property,
    300             schema_path=property,
    301         )

File ~/kf_pipelines_components/lib/python3.12/site-packages/jsonschema/validators.py:431, in create.<locals>.Validator.descend(self, instance, schema, path, schema_path, resolver)
    428     continue
    430 errors = validator(evolved, v, instance, schema) or ()
--> 431 for error in errors:
    432     # set details if not already set by the called fn
    433     error._set(
    434         validator=k,
    435         validator_value=v,
   (...)    438         type_checker=evolved.TYPE_CHECKER,
    439     )
    440     if k not in {"if", "$ref"}:

File ~/kf_pipelines_components/lib/python3.12/site-packages/jsonschema/_keywords.py:226, in format(validator, format, instance, schema)
    224 if validator.format_checker is not None:
    225     try:
--> 226         validator.format_checker.check(instance, format)
    227     except FormatError as error:
    228         yield ValidationError(error.message, cause=error.cause)

File ~/kf_pipelines_components/lib/python3.12/site-packages/jsonschema/_format.py:138, in FormatChecker.check(self, instance, format)
    136 result, cause = None, None
    137 try:
--> 138     result = func(instance)
    139 except raises as e:
    140     cause = e

Cell In[27], line 23, in custom_email_validation(instance)
     21 else:
     22   print("this is not string")
---> 23   raise RuntimeError("Some error in my custom logic due to unexpected error.")
     24 return False

RuntimeError: Some error in my custom logic due to unexpected error.

```

This should not happen in the first place as the schema restricts `my-field` to string only.

**Comment by Julian** [comment]
JSON, and thereby JSON Schema, has no notion of ordering of validators, all may be run in any order.
